File: go/979. Distribute Coins in Binary Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func distributeCoins(root *TreeNode) int {
	answer := 0
	var findAnswer func(root *TreeNode) int
	findAnswer = func(root *TreeNode) int {
		if root == nil {
			return 0
		}
		required := findAnswer(root.Left) + findAnswer(root.Right) + 1 - root.Val
		answer += int(math.Abs(float64(required)))
		return required
	}
	findAnswer(root)
	return answer
}

File: python/979. Distribute Coins in Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def distributeCoins(self, root: Optional[TreeNode]) -> int:
        answer = 0
        def find_answer(root):
            nonlocal answer
            if not root: return 0
            required = find_answer(root.left) + find_answer(root.right) + 1 - root.val
            answer += abs(required)
            return required
        find_answer(root)
        return answer